fhir-database-utils/src/main/java/com/ibm/fhir/database/utils/query/FromClause.java

```diff
@@ -8,46 +8,128 @@
 
     /**
-     * @param schemaName
-     * @param tableName
+     * Add a table without an alias. Tsk tsk.
+     * @param schemaName the schema qualifier for the table
+     * @param tableName the table name
      */
     public void addTable(String schemaName, String tableName) {
-        FromItemTable fit = new FromItemTable(schemaName, tableName);
-        items.add(fit);
+        TableRowSource fit = new TableRowSource(schemaName, tableName);
+        items.add(new FromItem(fit, null));
     }
 
+    /**
+     * Add a qualified schema.table with an alias.
+     * @param schemaName the schema qualifier for the table
+     * @param tableName the table name
+     * @param alias the alias to use in select expressions and join predicates
+     */
     public void addTable(String schemaName, String tableName, Alias alias) {
-        FromItemTable fit = new FromItemTable(schemaName, tableName, alias);
-        items.add(fit);
+        TableRowSource table = new TableRowSource(schemaName, tableName);
+        items.add(new FromItem(table, alias));
     }
 
+    /**
+     * Add a table to the FROM items list
+     * @param tableName
+     * @param alias
+     */
     public void addTable(String tableName, Alias alias) {
-        FromItemTable fit = new FromItemTable(null, tableName, alias);
-        items.add(fit);
+        TableRowSource table = new TableRowSource(tableName);
+        items.add(new FromItem(table, alias));
+    }
+
+    /**
+     * Add a table to the FROM items list
+     * @param tableName
+     * @param alias
+     */
+    public void addTable(String tableName) {
+        TableRowSource table = new TableRowSource(tableName);
+        items.add(new FromItem(table));
     }
 
     /**
      * Add the sub-query as an item in the from list
-     * 
+     *
      * @param sub
      * @param alias
      */
     public void addFrom(Select sub, Alias alias) {
-        FromItemSelect fis = new FromItemSelect(sub, alias);
-        items.add(fis);
+        SelectRowSource fis = new SelectRowSource(sub);
+        items.add(new FromItem(fis, alias));
     }
 
     @Override
     public String toString() {
-        return items.stream().map(FromItem::toString).collect(Collectors.joining(", "));
+        // Building the FROM string is a little more involved because of the
+        // potential to mix ANSI-style and traditional joins. The JOIN statements
+        // don't require , to separate them. We only need a comma if the next
+        // thing we're going to join is ANSI-style.
+        if (items.isEmpty()) {
+            throw new IllegalStateException("No FROM elements");
+        }
+
+        StringBuilder result = new StringBuilder();
+        FromItem element = items.get(0);
+        result.append(element.toString());
+        for (int i=1; i<items.size(); i++) {
+            FromItem nextElement = items.get(i);
+            if (nextElement.isAnsiJoin()) {
+                result.append(" "); // e.g. INNER JOIN ... ON ...
```

Comment:
Previously we had JDBC constants... should this and line 94 come from a constants file?

Reply:
I'm deferring constants refactor until we get rid of the old query builder. There's a lot of stuff which can get removed then.

Reply:
maybe we create a backlog issue then